### ISSUE TITLE
publish from dispached queue

### ISSUE THREAD

**Issue by kirillivanov** [ckrey/MQTT-Client-Framework#56]
Hi, next code will indefinitely lock main thread of the application. Any thoughts?
```
dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        [_mqttClient publishJson:json onTopic:topic]; // try to call  any other publish method include "AndWait"


});
```
Thread 1 and 8 are using the same managed context of core data
![screen shot 2015-10-14 at 19 10 27](https://cloud.githubusercontent.com/assets/903240/10490067/d0632c3a-72a7-11e5-9b5c-aa07543dde0f.png)

Apple Note:
You should not, therefore, initialize a context on one thread then pass it to a different thread. Instead, you should pass a reference to a persistent store coordinator and have the receiving thread/queue create a new context derived from that.

**Comment by ckrey** [comment]
The ManagedObjectContext for persistence is created when the MQTTSession is initialized.
Subsequent calls to any MQTTSession methods (such as publishJson) will use this context.

To let me understand what you are trying to achieve: Why do you put the publishJson into a dispatch_asynch? Publishing in MQTTClient is already asynchronous, you may call publish from the mainThread directly.

**Comment by kirillivanov** [comment]
For example someone want to use dispatched groups...

```
dispatch_group_t group = dispatch_group_create();

dispatch_group_async(group,dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^ {
    // block1
    [_mqttClient publishJson:json onTopic:topic1]; 
});

dispatch_group_async(group,dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^ {
    // block2
    [_mqttClient publishJson:json onTopic:topic2];
});

dispatch_group_notify(group,dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^ {
    // block3
    [_mqttClient publishJson:json onTopic:topic3]; 
});
```

**Comment by ckrey** [comment]
I see what you are doing. You run MQTTSession in the main thread (including the callbacks `handle event` etc) and do the publish in other threads.

This does not work (or was not taken care of by me). What will work is.
```
dispatch_group_async(group,dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^ {
    // block1
   ...
   [self performOnMainThread:@selector(publish:) withObject:... ]
   ...
});

- (void)publish:(... {
   [_mqttClient publishJson:....
}

```

**Comment by TiagoVeloso** [comment]
Is this planned to work?

I mean are you considering to support this? (sending messages from a dispatched queue?)

**Comment by ckrey** [comment]
I assume this is fixed with #61, please re-test
